docs/building-modules.md

```diff
@@ -123,8 +123,89 @@ individual methods in
 
 - [DOMException](https://developer.mozilla.org/en-US/docs/Web/API/DOMException)
 
+### libp2p
+
+Zinnia comes with a built-in libp2p node based on
+[rust-libp2p](https://github.com/libp2p/rust-libp2p). The node is shared by all
+Station Modules running on Zinnia. This way we can keep the number of open
+connections lower and avoid duplicate entries in routing tables.
+
+The initial version comes with a limited subset of features. We will be adding
+more features based on feedback from our users (Station Module builders).
+
+#### Networking stack
+
+- Transport: `tcp` using system DNS resolver
+- Multistream-select V1
+- Authentication: `noise` with `XX` handshake pattern using X25519 DH keys
+- Stream multiplexing: both `yamux` and `mplex`
+
+#### `Zinnia.dialProtocol(multiaddr, protoName)`
+
+Dial a remote peer identified by the `multiaddr` and open a new substream for
+the protocol identified by `protoName`. The peer address must include both the
+network address and peer id.
+
+```js
+const stream = Zinnia.dialProtocol(
+  // peer address + peer id
+  "/dns/example.com/tcp/3030/p2p/12D3KooWRH7asdJasdr136asXdasd723MasdasdX3qiasda8sasd",
+  // protocol name
+  "/ipfs/ping/1.0.0",
+);
+```
+
+**TBD: How should the Stream API look like?**
+
+Ideally, we want to use Web Streams:
```

Comment:
While this is ideal, how complex will it be, and how unstable is the tech? Would it be better for this stage of the project to use the simplest possible streams? Maybe iterators, event emitters, nodejs streams?

Reply:
Great comment!

I was thinking a lot about this topic over the weekend. Considering the complexity of integration with rust-libp2p, I decided to simplify the first version of libp2p API radically, remove all streaming and use buffers (`Uint8Array`) instead - both for the request and the response.

```ts
// exposed as Zinnia.requestProtocol
function requestProtocol(
  multiaddr: string,
  protoName: string,
  requestPayload: Uint8Array,
): Promise<Uint8Array>;
```

Reply:
yes! Maybe we could even implement ttfb measurements like this, if we add an option to drop after the first byte or so

Reply:
On the second thought, I'd prefer to return something else than `Uint8Array` to be forward compatible.

```ts
requestProtocol(
  multiaddr: string,
  protoName: string,
  requestPayload: Uint8Array,
): Promise<PeerResponse>;
```

Where `PeerResponse` can be modelled after Fetch API `Response` and provide a single method `arrayBuffer()`, or maybe implement an async iterator that will initially yield the entire response in one iteration.

Reply:
> Maybe we could even implement ttfb measurements like this, if we add an option to drop after the first byte or so

I'd like to implement the streaming of responses ASAP after we launch the first alpha. Streaming will allow you to measure TTFB without any additional Zinnia APIs.